src/OpenIddict.EntityFrameworkCore/Stores/OpenIddictEntityFrameworkCoreTokenStore.cs

```diff
@@ -626,55 +626,50 @@ public virtual IAsyncEnumerable<TToken> ListAsync(int? count, int? offset, Cance
     {
-        // Note: Entity Framework Core doesn't support set-based deletes, which prevents removing
-        // entities in a single command without having to retrieve and materialize them first.
-        // To work around this limitation, entities are manually listed and deleted using a batch logic.
-
         List<Exception>? exceptions = null;
 
-        async ValueTask<IDbContextTransaction?> CreateTransactionAsync()
-        {
-            // Note: transactions that specify an explicit isolation level are only supported by
-            // relational providers and trying to use them with a different provider results in
-            // an invalid operation exception being thrown at runtime. To prevent that, a manual
-            // check is made to ensure the underlying transaction manager is relational.
-            var manager = Context.Database.GetService<IDbContextTransactionManager>();
-            if (manager is IRelationalTransactionManager)
-            {
-                // Note: relational providers like Sqlite are known to lack proper support
-                // for repeatable read transactions. To ensure this method can be safely used
-                // with such providers, the database transaction is created in a try/catch block.
-                try
-                {
-                    return await Context.Database.BeginTransactionAsync(IsolationLevel.RepeatableRead, cancellationToken);
-                }
-
-                catch (Exception exception) when (!OpenIddictHelpers.IsFatal(exception))
-                {
-                    return null;
-                }
-            }
-
-            return null;
-        }
+        // Note: the Oracle MySQL provider doesn't support DateTimeOffset and is unable
+        // to create a SQL query with an expression calling DateTimeOffset.UtcDateTime.
+        // To work around this limitation, the threshold represented as a DateTimeOffset
+        // instance is manually converted to a UTC DateTime instance outside the query.
+        var date = threshold.UtcDateTime;
 
         // Note: to avoid sending too many queries, the maximum number of elements
         // that can be removed by a single call to PruneAsync() is deliberately limited.
         for (var index = 0; index < 1_000; index++)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
+#if SUPPORTS_BULK_DBSET_OPERATIONS
+            try
+            {
+                await (from token in Tokens
+                       where token.CreationDate < date
+                       where (token.Status != Statuses.Inactive && token.Status != Statuses.Valid) ||
+                             (token.Authorization != null && token.Authorization.Status != Statuses.Valid) ||
+                              token.ExpirationDate < DateTime.UtcNow
+                       orderby token.Id
+                       select token).Take(50_000).ExecuteDeleteAsync(cancellationToken);
```

Comment:
This will result in a full table scan.  The token Id (clustered index) is a Guid and a new one is just as likely to end up at the beginning of the table as the end.  This will halt a large system like ours (200 million tokens).  The 50k batch size will result in a table lock which will be a non-starter for some (including me) so can this be made configurable?

Reply:
> The 50k batch size will result in a table lock which will be a non-starter for some (including me) so can this be made configurable?

If we reduce the batch size, we need to make more passes to eliminate the same number of entries. Or we need to accept there may be many remaining entries after a single call to `PruneAsync()`.

I'm open to all suggestions: what would be a more appropriate value for you?

>  This will halt a large system like ours (200 million tokens).

For such demanding deployments, I'd recommend going DB-first and writing your own stores - e.g based on Dapper or ADO.NET directly - instead of using the EF Core stores and the default generated schema.

Reply:
Batches of 1_000 are generally comfortable, at 5_000 (in my experience) you'll start to experience lock escalation.  For sites that can't tolerate the downtime, the alternative is turning off pruning so leaving behind some tokens should be fine.

I rolled my own clean up script, but added an index on, and only looked at `ExpirationDate`.  You might want to consider that here as well.  With `CreationDate` and `Status` in the predicate, SqlServer will just scan the CX, and the oldest, most expired token might be the very last one.  That will create seizures in even a modest size system.

I'll look closer at stores in the new year once we're through our busy season, but our biggest challenge has been the number of tokens stored (3 every time we use a refresh token to get an access token) and the storage size, rather that EF performance.  I'm getting off an a tangent here, so I'll put together observations, suggestions and maybe even a PR or two in the new year.  Thanks for this project.

Reply:
As we have a datetime threshold, it could be great to have a limit to the batch size.